From 7d8040450de11d5268c5e6fe96dc31ef01e10f07 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Markus=20M=C3=BCtzel?= <markus.muetzel@gmx.de>
Date: Sun, 11 Jan 2026 19:57:00 +0100
Subject: [PATCH] flang-rt: Avoid duplicate definition of
 `std::__libcpp_verbose_abort`.

If a project depends on the Fortran runtime and on libc++, linking fails
because `std::__libcpp_verbose_abort` is defined in both libraries.

Avoid that duplicate definition by defining `_LIBCPP_VERBOSE_ABORT` before
including any C++ headers and by renaming that symbol in the Fortran runtime
to `flang_rt_verbose_abort`.
---
 cmake/modules/AddFlangRT.cmake |  6 ++++++
 lib/runtime/io-api-minimal.cpp |  7 +++----
 lib/runtime/io-api-minimal.h   | 12 ++++++++++++
 3 files changed, 21 insertions(+), 4 deletions(-)
 create mode 100644 lib/runtime/io-api-minimal.h

diff --git a/cmake/modules/AddFlangRT.cmake b/cmake/modules/AddFlangRT.cmake
index 923507764d69..e8b714693a8a 100644
--- a/cmake/modules/AddFlangRT.cmake
+++ b/cmake/modules/AddFlangRT.cmake
@@ -317,6 +317,12 @@ function (add_flangrt_library name)
           "$<$<COMPILE_LANGUAGE:CXX>:-Wp,-U_LIBCPP_ENABLE_ASSERTIONS>")
     endif ()
 
+    # Use own function in place of `std::__libcpp_verbose_abort` in the Fortran
+    # runtime to avoid an unwanted dependency on the symbol provided by libc++.
+    target_compile_options(${tgtname} PRIVATE
+      $<$<COMPILE_LANGUAGE:CXX>:"-D_LIBCPP_VERBOSE_ABORT(...)=flang_rt_verbose_abort(__VA_ARGS__)" -include "${FLANG_RT_SOURCE_DIR}/lib/runtime/io-api-minimal.h">
+      )
+
     # Non-GTest unittests depend on LLVMSupport
     if (ARG_LINK_TO_LLVM)
       if (LLVM_LINK_LLVM_DYLIB)
diff --git a/lib/runtime/io-api-minimal.cpp b/lib/runtime/io-api-minimal.cpp
index fdf7183ed517..dad1c1e901e1 100644
--- a/lib/runtime/io-api-minimal.cpp
+++ b/lib/runtime/io-api-minimal.cpp
@@ -147,11 +147,10 @@ bool IODEF(OutputLogical)(Cookie cookie, bool truth) {
 } // namespace Fortran::runtime::io
 
 #if defined(_LIBCPP_VERBOSE_ABORT)
-// Provide own definition for `std::__libcpp_verbose_abort` to avoid dependency
-// on the version provided by libc++.
+// Provide function that is used in place of `std::__libcpp_verbose_abort` in
+// the Fortran runtime to avoid dependency on the symbol provided by libc++.
 
-void std::__libcpp_verbose_abort(char const *format, ...) noexcept(
-    noexcept(std::__libcpp_verbose_abort(""))) {
+void flang_rt_verbose_abort(char const *format, ...) {
   va_list list;
   va_start(list, format);
   std::vfprintf(stderr, format, list);
diff --git a/lib/runtime/io-api-minimal.h b/lib/runtime/io-api-minimal.h
new file mode 100644
index 000000000000..5ea5c2ea56fc
--- /dev/null
+++ b/lib/runtime/io-api-minimal.h
@@ -0,0 +1,12 @@
+//===-- lib/runtime/io-api-minimal.h --------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// Declare function that is used in place of `std::__libcpp_verbose_abort` in
+// the Fortran runtime to avoid dependency on the symbol provided by libc++.
+
+void flang_rt_verbose_abort(char const *format, ...);
-- 
2.51.0.windows.2

